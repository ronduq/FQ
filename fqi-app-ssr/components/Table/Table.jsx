import React from 'react';
import classNames from 'classnames';
import get from 'lodash.get';
import LinkButton from '../../components/LinkButton/LinkButton';
import Icon from '../Icon/Icon';
import Tooltip from '../../hocs/TooltipContainer';
import s from './Table.module.scss';
import t from '../Icon/Icon.module.scss';

const Table = ({
  sortRanks, 
  retailersData, 
  noBorders, 
  labels, 
  selected, 
  direction, 
  ranks, 
  type, 
  produceId, 
  location,
}) => {
  const selectedCriteria = selected;
  const categories = {
    quality: {
      criteria: 'quality',
      id: 'tableHeaderQualityButton',
      firstOrder: 'bottomToTop',
      secondOrder: 'topToBottom',
      champion: 'qualityChampion',
    },
    value: {
      criteria: 'value',
      id: 'tableHeaderValueButton',
      firstOrder: 'topToBottom',
      secondOrder: 'bottomToTop',
      champion: 'valueChampion',
    },
    perception: {
      criteria: 'perception',
      id: 'tableHeaderPerceptionButton',
      firstOrder: 'topToBottom',
      secondOrder: 'bottomToTop',
      champion: 'perceptionChampion',
    },
  }

  const selectCriteria = (criteria) => {
    const selectedCriteria = selected;
    if (selectedCriteria === criteria && direction === 'topToBottom') {
      sortRanks(criteria, 'bottomToTop', type);
    } else if (selectedCriteria === criteria && direction === 'bottomToTop') {
      sortRanks(criteria, 'topToBottom', type);
    } else {
      if (criteria === 'quality'){
        sortRanks(criteria, 'topToBottom', type);
      } else {
        sortRanks(criteria, 'bottomToTop', type);
      }
    }
  }

  return (
    <table className={classNames(s.table, { [s.noBorders]: noBorders})}  role="grid" aria-readonly="true">
      <caption>Retailers table sorted by {selectedCriteria}, {direction === 'topToBottom' && <span>ascending</span>} {direction === 'bottomToTop' && <span>descending </span>} </caption>
      <thead>
        <tr className={s.headerRow}>
          <th className={classNames(s.tableHeader, s.cell)} scope="col">{labels.retailer}</th>
          {categories && Object.keys(categories).map((category, index) => {
            return (
              <th 
                className={classNames(s.tableHeader, s.cell, s.filterButton,{[s.cellHighlight] : selectedCriteria === categories[category].criteria})} 
                scope="col" 
                id={categories[category].id}
                key={index}
              >
                <Tooltip 
                  side='left' 
                  size='small'
                  bubblePosition='left'
                  id={categories[category].criteria}
                >
                <button onClick={() => { selectCriteria(categories[category].criteria)} } >
                  {labels[category]}
                  <div className={s.arrowsContainer}>
                    {(selectedCriteria !== categories[category].criteria ||
                    (selectedCriteria === categories[category].criteria && direction !== categories[category].secondOrder)) && (
                      <span className={classNames(s.arrow, s.arrowUp)}>
                        <Icon className={classNames(t.iconXSmall, t.iconBlack)} iconName={`chevron`} />
                      </span>
                    )}
                    {(selectedCriteria !== categories[category].criteria ||
                    (selectedCriteria === categories[category].criteria && direction !== categories[category].firstOrder)) && (
                      <span className={classNames(s.arrow, s.arrowDown)}>
                        <Icon className={classNames(t.iconXSmall, t.iconBlack)} iconName={`chevron`} />
                      </span>
                    )}
                  </div>
                </button>
                </Tooltip>
                
              </th>
            )
          })}
          <th></th>
        </tr>
      </thead>

      <tbody>
        {Object.keys(ranks).map((item,index) => {
          const retailer = ranks[item];
          const { retailerCode } = retailer;

          let linkHref = `/retailer-profile?location=${location}&retailer=${retailerCode}`;
          let linkAs = `/${location}/retailers/${retailerCode}`
          let buttonLabel = labels.button_label;

          if (produceId) {
            linkHref = `/produce-profile?location=${location}&produce=${produceId}&retailer=${retailerCode}`
            linkAs = `/${location}/produce/${produceId}/${retailerCode}/`
            buttonLabel = labels.produce_button_label;
          }

          return (
            <tr className={s.bodyRow} key={index}>
              <td 
                className={classNames(s.cell, s.cellImage)}
                data-title={labels.retailer}
              >
                <img className={s.logo} src={`${get(retailersData.itemsLogos,[retailerCode])}`} alt={retailerCode}/>
              </td>
              {Object.keys(categories).map((category,idx) => {
                return(
                  <td 
                    className={classNames(
                      s.largeText, 
                      s.cell, 
                      {[s.cellHighlight] : selectedCriteria === categories[category].criteria},
                      {[s.cellChampion] : retailer[categories[category].champion]},
                    )}
                    data-title={labels[categories[category].criteria]}
                    key={idx}
                  >
                    {retailer[categories[category].criteria]}{(categories[category].criteria === 'quality') && <span>%</span>} {
                    retailer[categories[category].champion] && 
                    (<Icon className={classNames(t.iconLarge, t.iconTeal, s.iconRosette)} iconName={`rosette`} aria-hidden={true}/>)} 
                  </td>
                )
              })}
              
              <td className={classNames(s.largeText, s.cell)}>
                <LinkButton label={buttonLabel} href={linkHref} as={linkAs} />
              </td>
            </tr>
          )
        })}
        
      </tbody>

    </table>
  )}


export default Table;