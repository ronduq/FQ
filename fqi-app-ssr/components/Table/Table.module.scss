@import '../../style/settings/colours';
@import '../../style/settings/fonts';
@import '../../style/settings/breakpoints';

.table {
  background: white;
  width: 100%;
  border: 1px solid $colour-grey-medium;
  border-collapse: separate;
  @include font-family('proxima-regular');
  &.noBorders {
    border: 0;
  }
  td {
    border-bottom: 1px solid $colour-grey-medium;
  }
  caption {
    opacity: 0;
    position: absolute;
    left: -99999999px;
  }

  @include respond-to('mobile') {
    display: block;
    & tbody, thead {
      display: block;
    }
  }
};

.headerRow {
  position:relative;
  @include font-family('proxima-light');

  @include respond-to('mobile') {
    display: none;
  }
  .cellHighlight {
    border-top: 1px solid $colour-grey-medium;
    .noBorders & {
      border: 0;
    }
  }
  button {
    background: none;
    border: none;
    appearance: none;
    @include font-size('small');
    cursor: pointer;
    width: 100%;
  }
  & th:first-child {
    text-align: left;
    padding-left: 20px;
  }
}

.bodyRow {
  position:relative;
  z-index: 1;
  border-bottom: 1px solid $colour-grey-medium;
  &:first-child {
    text-align: left;
    .cell {
      box-shadow: 0 10px 4px -9px rgba(126,126,126, 0.7) inset;
      @include respond-to('mobile') {
        box-shadow: none;
      }
    }
  }
  
  @include respond-to('mobile') {
    display: block;
    border-bottom: 2px solid $colour-grey-dark;

    td { 
      display: block;
      border: none;
      border-bottom: 1px solid $colour-grey-medium; 
      position: relative;
      padding-left: 50%; 
      padding-right: 16px;
      text-align: right;
      @include font-size('medium');
      @include font-family('proxima-bold');

      &:nth-of-type(1) {
        padding-top: 16px;
        padding-bottom: 16px;
      }
    }
    
    td:before { 
      position: absolute;
      left: 16px;
      width: 35%; 
      padding-right: 16px; 
      white-space: nowrap;
      text-align: left;
      @include font-size('medium');
      @include font-family('proxima-light');
    }
    td:nth-of-type(1):before { content: attr(data-title); }
    td:nth-of-type(2):before { content: attr(data-title); }
    td:nth-of-type(3):before { content: attr(data-title); }
    td:nth-of-type(4):before { content: attr(data-title); }
  }
}

.cell {
  padding: 16px 0;
  text-align: center;
  box-sizing: border-box;
  vertical-align: middle;
}

.cellImage {
  padding: 0px 0px 0px 16px;
  text-align: left;
}

.cellChampion {
  color: $colour-main;
  @include respond-to('mobile') {
    color: $colour-grey-dark;
  }
}

.largeText {
  @include font-family('proxima-regular');
  @include font-size('large');
}

.cellHighlight {
  background: $colour-grey-light;
  @include font-family('proxima-bold');
  @include respond-to('mobile') {
    background: none;
    @include font-family('proxima-regular');
  }

  button {
    @include font-family('proxima-bold');
  }

}

.filterButton {
  cursor: pointer;
}

.logo {
  height: 30px;
  position: relative;
  display: block;
  
  @include respond-to('mobile') {
   margin-left: auto;
   height: 20px;
  }
}

.arrowsContainer {
  position: relative;
  vertical-align: top;
  display: inline-block;
  margin-left: 5px;
  @include respond-to('tablet') {
    display: none;
  }
}
.arrow {
  position: absolute;
  transition: ease-in-out transform .1s;
}
.arrowUp {
  transform: translateY(-4px) rotate(180deg);
  .filterButton:hover &{
    transform: translateY(-6px) rotate(180deg);
  }
}
.arrowDown {
  transform: translateY(4px);
  .filterButton:hover &{
    transform: translateY(6px);
  }
}

.iconRosette {
  vertical-align: middle;
  transform: translateY(-2px);
  margin-left: -6px;
}