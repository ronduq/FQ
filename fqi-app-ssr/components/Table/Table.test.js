/* eslint-env jest */

import { mount, shallow } from 'enzyme'
import React from 'react'
import Table from './Table'
import { Provider } from 'react-redux';

describe('With Snapshot Testing', () => {

  const props = {
    // retailerRanksData: {
      ranks: {
        "0": { 
          retailerCode: "whole-foods",
          quality: 95,
          value: 3,
          perception: 6,
          qualityChampion: true,
          valueChampion: false,
          perceptionChampion: false 
        },
        "1": { 
          retailerCode: "walmart",
          quality: 93,
          value: 1,
          perception: 2,
          qualityChampion: false,
          valueChampion: true,
          perceptionChampion: false 
        },
        "2": { 
          retailerCode: "costco",
          quality: 90,
          value: 2,
          perception: 3,
          qualityChampion: false,
          valueChampion: false,
          perceptionChampion: false 
        },
        "3": { 
          retailerCode: "amazon-fresh",
          quality: 89,
          value: 5,
          perception: 4,
          qualityChampion: false,
          valueChampion: false,
          perceptionChampion: false 
        },
        "4": { 
          retailerCode: "gelsons",
          quality: 83,
          value: 6,
          perception: 5,
          qualityChampion: false,
          valueChampion: false,
          perceptionChampion: false 
        },
        "5": { 
          retailerCode: "aldi",
          quality: 76,
          value: 4,
          perception: 1,
          qualityChampion: false,
          valueChampion: false,
          perceptionChampion: true 
        },
      },
      labels: {
        retailer: 'Retailer',
        quality: 'Quality Score',
        value: 'Value Rank',
        perception: 'Perception Rank',
      },
      selected: 'quality',
      direction: 'topToBottom',
    // },
    retailersData: {
      name: 'Retailers',
      items: {
        '0': {id: 'costco', selected: true, isChild: false},
        '1': {id: 'walmart', selected: true, isChild: false},
        '2': {id: 'whole-foods', selected: true, isChild: false},
        '3': {id: 'aldi', selected: true, isChild: false},
        '4': {id: 'bristol-farms', selected: true, isChild: false},
        '5': {id: 'amazon-fresh', selected: true, isChild: false},
        '6': {id: 'gelsons', selected: true, isChild: false},
        '7': {id: 'ralphs', selected: true, isChild: false},
        '8': {id: 'stop-n-shop', selected: true, isChild: false},
        '9': {id: 'target', selected: true, isChild: false},
        '10': {id: 'trader-joes', selected: true, isChild: false},
        '11': {id: 'wegmans', selected: true, isChild: false},
      },
      allItemsSelected: true,
      itemsLabels: {},
      itemsLogos: {},
      
    }
  }

  const selectCriteria = jest.fn();
  const sortRanks = jest.fn();

  const categories = {
    quality: {
      criteria: 'quality',
      id: 'tableHeaderQualityButton',
      firstOrder: 'bottomToTop',
      secondOrder: 'topToBottom',
      champion: 'qualityChampion',
    },
    value: {
      criteria: 'value',
      id: 'tableHeaderValueButton',
      firstOrder: 'topToBottom',
      secondOrder: 'bottomToTop',
      champion: 'valueChampion',
    },
    perception: {
      criteria: 'perception',
      id: 'tableHeaderPerceptionButton',
      firstOrder: 'topToBottom',
      secondOrder: 'bottomToTop',
      champion: 'perceptionChampion',
    },
  }

  const fakeStore = {
    dispatch: jest.fn(),
    getState: () => ({
      content: {
        tooltips:{
          trend:{}
        }
      }
    }),
    subscribe: jest.fn()
  };
  
  it('Table snapshot"', () => {
    const wrapper = mount(
      <Provider store={fakeStore}>
        <Table {...props}/>
      </Provider>);
    
    expect(wrapper).toMatchSnapshot()
  });

  it('Sorts top to bottom', () => {
    const wrapper = mount(
      <Provider store={fakeStore}>
        <Table {...props} sortRanks={sortRanks} selected={'quality'} direction={'topToBottom'}/>
      </Provider>);
    const qualityButton = wrapper.find('#tableHeaderQualityButton button');
    qualityButton.simulate('click');
    expect(sortRanks).toHaveBeenCalledTimes(1);
  });

  it('Sorts bottom to top', () => {
    const wrapper = mount(
      <Provider store={fakeStore}>
        <Table {...props} sortRanks={sortRanks} selected={'value'} direction={'bottomToTop'}/>
      </Provider>);
    const qualityButton = wrapper.find('#tableHeaderValueButton button');
    qualityButton.simulate('click');
    expect(sortRanks).toHaveBeenCalledTimes(2);
  });

  it('Sorts top to bottom on quality', () => {
    const wrapper = mount(
      <Provider store={fakeStore}>
        <Table {...props} sortRanks={sortRanks} selected={'value'}/>
      </Provider>);
    const qualityButton = wrapper.find('#tableHeaderQualityButton button');
    qualityButton.simulate('click');
    expect(sortRanks).toHaveBeenCalledTimes(3);
  });

  it('Sorts top to bottom on quality', () => {

    const wrapper = mount(
      <Provider store={fakeStore}>
        <Table {...props} sortRanks={sortRanks} selected={'quality'}/>
      </Provider>);
    const qualityButton = wrapper.find('#tableHeaderPerceptionButton button');
    qualityButton.simulate('click');
    expect(sortRanks).toHaveBeenCalledTimes(4);
  });
})