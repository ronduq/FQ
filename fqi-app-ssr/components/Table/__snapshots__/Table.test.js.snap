// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`With Snapshot Testing Table snapshot" 1`] = `ReactWrapper {}`;
